src/main/java/org/jenkinsci/plugins/workflow/log/LogStorage.java

```diff
@@ -0,0 +1,129 @@
+     */
+    @Nonnull BuildListener overallListener() throws IOException, InterruptedException;
+
+    /**
+     * Provides an alternate way of emitting output from a node (such as a step).
+     * Currently not closed at the end of a step (TBD if this would be useful).
+     * @param node a running node
+     * @return a (remotable) task listener; do not bother overriding anything except {@link TaskListener#getLogger}
+     * @see StepContext#get
+     */
+    @Nonnull TaskListener nodeListener(@Nonnull FlowNode node) throws IOException, InterruptedException;
+
+    /**
+     * Provides an alternate way of retrieving output from a build.
+     * <p>In an {@link AnnotatedLargeText#writeHtmlTo} override, {@link ConsoleAnnotationOutputStream#eol}
+     * should apply {@link #startStep} and {@link #endStep} to delineate blocks contributed by steps.
+     * (Also see {@link ConsoleAnnotators}.)
+     * @param complete if true, we claim to be serving the complete log for a build,
+     *                  so implementations should be sure to retrieve final log lines
+     * @return a log
+     */
+    @Nonnull AnnotatedLargeText<FlowExecutionOwner.Executable> overallLog(@Nonnull FlowExecutionOwner.Executable build, boolean complete);
+
+    /**
+     * Introduces an HTML block with a {@code pipeline-node-<ID>} CSS class based on {@link FlowNode#getId}.
+     * @see #endStep
+     * @see #overallLog
+     */
+    static @Nonnull String startStep(@Nonnull String id) {
+        return "<span class=\"pipeline-node-" + id + "\">";
+    }
+
+    /**
+     * Closes an HTML step block.
+     * @see #startStep
+     * @see #overallLog
+     */
+    static @Nonnull String endStep() {
+        return "</span>";
+    }
+
+    /**
+     * Provides an alternate way of retrieving output from a build.
+     * @param node a running node
+     * @param complete if true, we claim to be serving the complete log for a node,
+     *                  so implementations should be sure to retrieve final log lines
+     * @return a log for this just this node
+     * @see LogAction
+     */
+     @Nonnull AnnotatedLargeText<FlowNode> stepLog(@Nonnull FlowNode node, boolean complete);
+
+    /**
+     * Gets the available log storage method for a given build.
+     * @param b a build about to start
+     * @return the mechanism for handling this build, including any necessary fallback
+     * @see LogStorageFactory
+     */
+    static @Nonnull LogStorage of(@Nonnull FlowExecutionOwner b) {
+        try {
+            for (LogStorageFactory factory : ExtensionList.lookup(LogStorageFactory.class)) {
+                LogStorage storage = factory.forBuild(b);
+                if (storage != null) {
+                    // TODO consider saving this decision
+                    return storage;
+                }
+            }
+            // Similar to Run.getLogFile, but not supporting gzip:
+            return StreamLogStorage.forFile(new File(b.getRootDir(), "log"), b);
```

Comment:
Yes, after discussion with @jglick and @dwnusbaum the best path is to keep something with similar performance scaling to the older file-per-step logging -- we get to keep good seek time so every single step where we view logs in BlueOcean and StageView, only results in reading *that* part of the log, rather than the entirety of the log (potentially hundreds of MB or more). 

It was *almost* a wash vs trying to fix the streaming implementation to support fast seeks without reading the entire log, but we opted to not go that path at this time to allow more time to work on that.

Reply:
After thinking it over a bit more, I am optimistic that a single-file implementation with a rangeset side file may not be that complicated. I will take a shot at it, since that would be a better outcome.

Reply:
@jglick You're welcome to, but after thinking on it a bit myself I think we *really* should rely on having potentially >1 implementation on-master, in case we discover a performance failure mode with this (as noted, rangesets are weak with interleaved parallel branch content, and I can name at least one user who does that with massively parallel branches).